terraform/ec2/asg/main.tf

```
@@ -0,0 +1,200 @@
+provider "aws" {}
+
+resource "aws_default_vpc" "default" {}
+
+data "aws_subnets" "default_subnets" {
+  filter {
+    name   = "vpc-id"
+    values = [aws_default_vpc.default.id]
+  }
+}
+
+resource "tls_private_key" "ssh_key" {
+  algorithm = "RSA"
+  rsa_bits = 4096
+}
+
+resource "aws_key_pair" "aws_ssh_key" {
+  key_name = "instance_key-${var.test_id}"
+  public_key = tls_private_key.ssh_key.public_key_openssh
+}
+
+locals {
+  ssh_key_name        = aws_key_pair.aws_ssh_key.key_name
+  private_key_content = tls_private_key.ssh_key.private_key_pem
+}
+
+data "aws_ami" "ami" {
+  owners = ["amazon"]
+  most_recent      = true
+  filter {
+    name   = "name"
+    values = ["al20*-ami-minimal-*-x86_64"]
+  }
+  filter {
+    name   = "state"
+    values = ["available"]
+  }
+  filter {
+    name   = "architecture"
+    values = ["x86_64"]
+  }
+  filter {
+    name   = "image-type"
+    values = ["machine"]
+  }
+
+  filter {
+    name   = "root-device-name"
+    values = ["/dev/xvda"]
+  }
+
+  filter {
+    name   = "root-device-type"
+    values = ["ebs"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+}
+
+resource "aws_launch_configuration" "launch_configuration" {
+  image_id      = data.aws_ami.ami.id
+  instance_type = "t3.micro"
+  key_name = local.ssh_key_name
+  associate_public_ip_address = true
+  iam_instance_profile = "APP_SIGNALS_EC2_TEST_ROLE"
+  security_groups = [aws_default_vpc.default.default_security_group_id]
+
+  user_data = <<-EOF
+    #!/bin/bash
+    mkdir test222
+    # Make the Terraform fail if any step throws an error
+    set -o errexit
+    # Install Java 11 and wget
+    sudo yum install wget java-11-amazon-corretto -y
+
+    # Copy in CW Agent configuration
+    agent_config='${replace(replace(file("./amazon-cloudwatch-agent.json"), "/\\s+/", ""), "$REGION", var.aws_region)}'
+    echo $agent_config > amazon-cloudwatch-agent.json
+
+    # Get and run CW agent rpm
+    ${var.get_cw_agent_rpm_command}
+    sudo rpm -U ./cw-agent.rpm
+    sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -m ec2 -s -c file:./amazon-cloudwatch-agent.json
+
+    # Get ADOT
+    ${var.get_adot_jar_command}
+
+    # Get and run the sample application with configuration
+    aws s3 cp ${var.sample_app_jar} ./main-service.jar
+
+    JAVA_TOOL_OPTIONS=' -javaagent:/adot.jar' \
+    OTEL_METRICS_EXPORTER=none \
+    OTEL_SMP_ENABLED=true \
+    OTEL_AWS_SMP_EXPORTER_ENDPOINT=http://localhost:4315 \
+    OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://localhost:4315 \
+    OTEL_RESOURCE_ATTRIBUTES=aws.hostedin.environment=EC2,service.name=sample-application-${var.test_id} \
+    nohup java -jar main-service.jar &> nohup.out &
+
+    # The application needs time to come up and reach a steady state, this should not take longer than 30 seconds
+    sleep 30
+    EOF
+}
+
+resource "aws_autoscaling_group" "asg" {
+  name                 = "asg-${var.test_id}"
```

We should have a better name in case we want to extend our test cases in the future, e.g.
```suggestion
  name                 = "ec2-single-asg-${var.test_id}"
```

---

Done